### ISSUE TITLE
Wrong urls redirected from downloading files when using base_url

### ISSUE THREAD

**Issue by w1ndy** [jupyter/notebook#1448]
When I tried to download a file from a notebook server, I'm getting redirected to '/files/...' url rather than '/some_configured_base_url/files/...' and hit a 404 error. However if I manually edit the url to '/some_configured_base_url/files/...', it'll work just fine.

**Comment by Carreau** [comment]
Can you give an example notebook ?

**Comment by w1ndy** [comment]
@Carreau 

Sorry but the notebook server I'm using is hosted inside the company's network. However I can provide  a Dockerfile which I'm using for deploying the server:

```dockerfile
FROM jupyter/notebook

ENV DEPS r-base r-base-dev r-recommended dvipng ghostscript poppler-utils libfreetype6-dev libpng12-dev libpng12-dev libjpeg-dev
ENV PACKAGES numpy requests pillow pymongo rpy2 scipy beautifulsoup4 networkx matplotlib scikit-learn

RUN sed -i 's%archive.ubuntu.com%mirrors.ustc.edu.cn/ubuntu%g' /etc/apt/sources.list && \
    echo '\ndeb http://mirrors.tuna.tsinghua.edu.cn/CRAN/bin/linux/ubuntu trusty/\n' >> /etc/apt/sources.list && \
    apt-key adv --keyserver keyserver.ubuntu.com --recv-keys E084DAB9

RUN mkdir -p ~/.pip && \
    echo '[global]\nindex-url = https://pypi.mirrors.ustc.edu.cn/simple' > ~/.pip/pip.conf

RUN apt-get update && \
    apt-get install -y $DEPS && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

RUN for pkg in $PACKAGES; do pip2 --no-cache-dir install $pkg; done
RUN for pkg in $PACKAGES; do pip3 --no-cache-dir install $pkg; done

RUN echo "c.NotebookApp.base_url = '/ipython/'" >> /root/.jupyter/jupyter_notebook_config.py
```
as well as the ``docker-compose.yml`` I'm using for starting up the server:
```yaml
jupyter:
    restart: always
    image: 10.76.2.13:5001/jupyter
    ports:
        - 9088:8888
    volumes:
        - /docker/jupyter:/notebooks
```
and nginx configuration for ``/ipython/`` location:
```
    location ^~ /ipython {
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Host $http_host;

        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        proxy_pass http://127.0.0.1:9088;
        proxy_redirect off;
    }
```

I believe it's a clean setup: the only thing I changed is the base_url. Could you please test with the configuration above? Thanks.

**Comment by Carreau** [comment]
Oh, sorry, I'm more interested in the content of the notebook file itself, just for a demo, like if your markdown link `[myfile](mifile.pdf)` or `[myfile](/myfile.pdf)`, or did I misunderstand and you download from the dashboard, or is it actually `file`> `download` ?

**Comment by w1ndy** [comment]
I wrote a small python script that retrieves data and save it to a JSON file in the same directory where the script is located. The issue occurred when I'm trying to download the JSON file directly from the tree view of the notebook.

**Comment by w1ndy** [comment]
![unti](https://cloud.githubusercontent.com/assets/652068/15273148/f33cd274-1ac2-11e6-8ac7-b59ba37b31c1.jpg)
This is the download I meant, it's redirecting me to a 404 page when ``base_url`` is enabled.

**Comment by w1ndy** [comment]
See https://github.com/jupyter/notebook/blob/master/notebook/static/tree/js/notebooklist.js#L888 , I believe
```javascript
window.open(utils.url_path_join('/files', item_path) + '?download=1');
```
should be 
```javascript
window.open(utils.url_path_join(this.base_url, '/files', item_path) + '?download=1');
```

And ``TODO`` on line 881 seems pretty easy to fix too:
```javascript
// TODO(nhdaly): Support download multiple items at once.
```
perhaps just one quick loop will do?

**Comment by takluyver** [comment]
@w1ndy that fix looks right to me, do you want to make a pull request?

Downloading multiple files might be a bit trickier, because I don't think the JS lets us trigger more than one download from one click. Usually the way things like webmail get round that is to let you download a zip file of the files you selected. We can do the same, but that's a separate issue from fixing this.

**Comment by minrk** [comment]
Fixed by #1916
